Enable support for injecting complex extra vars

philipsd6 opened the pull request:
##### SUMMARY
Until now, it's only been possible to inject flat strings with Jinja templating into extra vars from Custom Credential Types. However, it's quite desirable to be able to pass a data structure containing authentication information for certain collections/modules. For example, the [Infoblox collection](https://docs.ansible.com/ansible/latest/scenario_guides/guide_infoblox.html#credentials-and-authenticating) accepts authentication under the `nios_provider` key.

This PR enables that data structure to be created by an injector config.

related #1965 #13070

<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New or Enhanced Feature

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - Other

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
PYTHON=python3 make VERSION
awx: 21.9.1.dev32+g4af80777dd
```


##### ADDITIONAL INFORMATION
This PR also makes it possible to template the keys of an injector configuration, as requested in #13070.

For example, given an input configuration providing values for `environment`, `host`, `apikey`, `username`, it's possible to use an extra_vars injector configuration of:
```yaml
"{{ environment }}_auth":
  host: "{{ host }}"
  apikey: "{{ apikey }}"
  username: "{{ username }}"
  verify_ssl: "{{ verify_ssl }}"
```
resulting in the following extra var available to the playbooks:
```yaml
prod_auth:
  host: http://production.server
  apikey: 5hfthisj4hisj5hgnotj6hrealgj6
  username: philipsd6
  verify_ssl: True
```
This would lay the groundwork for allowing multiple credentials of the _same_ credential type, as another credential might supply values for a different environment.
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

philipsd6 committed:
Add function to walk the extra_vars and render the results

philipsd6 committed:
Render keys while walking extra_vars in addition to values

philipsd6 commented:
I'm opening this PR to generate some discussion -- I'll update with tests and documentation if this passes muster.

AlanCoding commented:
Let's get your example documented in `docs/credentials/custom_credential_types.md`

philipsd6 commented:
> Let's get your example documented in `docs/credentials/custom_credential_types.md`

I'll document the major part of this PR, but not the ability to template the key (yet) as that needs some discussion.

philipsd6 committed:
Update documentation to include subkey injection

AlanCoding commented on awx/main/models/credential/__init__.py:
```diff
@@ -540,6 +545,7 @@ def build_extra_vars_file(vars, private_dir):
                 os.chmod(path, stat.S_IRUSR)
                 return path
 
+            extra_vars = render_extravars(self.injectors.get('extra_vars', {}))
```
Should this be a call to `build_extra_vars` instead?

philipsd6 replied:
YES it SHOULD! My bad, I missed that when I refactored my code...

AlanCoding commented on awx/main/models/credential/__init__.py:
```diff
@@ -528,9 +528,14 @@ class TowerNamespace:
-                extra_vars[var_name] = sandbox_env.from_string(tmpl).render(**namespace)
+            def build_extra_vars(node):
+                if isinstance(node, dict):
+                    return {sandbox_env.from_string(k).render(**namespace):
```
my 2 cents, I'd rather call `build_extra_vars(k)` as opposed to `sandbox_env.from_string(k).render(**namespace)`. If we're recursive anyway, might as well own it.

shanemcd commented:
This is awesome 😄 Before we merge this we'll need to get some tests added. Check out `test_tasks.py` - there's a lot of examples of tests calling `inject_credential`.

philipsd6 committed:
Add two tests for nested and templated extra_vars keys

philipsd6 commented:
Added two tests for nested and templated extra vars keys, but still TBD: Updating the jsonschema for the `CredentialTypeInjectorField` model (the tests will fail until this is done.)

john-westcott-iv commented:
@philipsd6 if you are all set with your changes can you convert this from a draft PR to an opened PR?

philipsd6 commented:
> @philipsd6 if you are all set with your changes can you convert this from a draft PR to an opened PR?

I don't think it's ready yet -- I'm struggling with the jsonschema that will allow recursively nested injector vars. Currently in my test env, I'm still getting this error when trying to create a nested extra_vars injector:

    OrderedDict([('nested_auth', OrderedDict([('username', '{{ username }}'), ('password', '{{ password }}')]))]) is not valid under any of the given schemas

philipsd6 committed:
Add recursive properties to injectors jsonschema for extra_vars

philipsd6 commented:
I updated `awx/main/fields.py:CredentialTypeInjectorField` to try to make the `extra_vars` property recursive, but while the schema is ok with an online validator: https://www.jsonschemavalidator.net/s/Yr7ExLBf it still fails in AWX when trying to update a credential with a nested extra_var injector with:

    TypeError: Can't compile non template nodes
    10.0.2.100 PATCH /api/v2/credential_types/28/ - HTTP/1.1 500

philipsd6 committed:
Add functions to support recursive validation for extra_vars

philipsd6 commented:
> I updated `awx/main/fields.py:CredentialTypeInjectorField` to try to make the `extra_vars` property recursive, but while the schema is ok with an online validator: https://www.jsonschemavalidator.net/s/Yr7ExLBf it still fails in AWX when trying to update a credential with a nested extra_var injector with:
> 
> ```
> TypeError: Can't compile non template nodes
> 10.0.2.100 PATCH /api/v2/credential_types/28/ - HTTP/1.1 500
> ```
My bad, the jsonschema update worked fine, it was the validator that was failing on the nested part. Fixed with latest commit.

philipsd6 committed:
Expand pattern to support use of Jinja2 block delimiters

philipsd6 commented on awx/main/fields.py:
```diff
@@ -790,13 +790,11 @@ def schema(self, model_instance):
                     'patternProperties': {
-                        # http://docs.ansible.com/ansible/playbooks_variables.html#what-makes-a-valid-variable-name
-                        '^[a-zA-Z_]+[a-zA-Z0-9_]*$': {'type': 'string'},
+                        r'^(?:(?:{(?:{|%)[^{}]*?(?:%|})})|(?:[a-zA-Z_]+[a-zA-Z0-9_]*)+)+$': {"anyOf": [{'type': 'string'}, {'$ref': '#/properties/extra_vars'}]}
```
This regex is ridiculous, but the only alternative I can see is to just make it `r'^.+$'`

AlanCoding commented on awx/main/fields.py:
```diff
@@ -858,27 +856,41 @@ def __str__(self):
+                raise django_exceptions.ValidationError(_('Encountered unsafe code execution: {}').format(e))
+            except TemplateSyntaxError as e:
+                raise django_exceptions.ValidationError(
+                    _('Syntax error rendering template for {sub_key} inside of {type} ({error_msg})').format(sub_key=key, type=type_, error_msg=e),
```
I'm getting a server error due to the fact that `key` is undefined here.

philipsd6 replied:
yup, valid -- and not only that, so is `type_`. I'll have to revise a little deeper after closer inspection. Thanks for the review.

philipsd6 committed:
Account for validation context

AlanCoding commented on awx/main/fields.py:
```diff
@@ -858,27 +856,41 @@ def __str__(self):
+
+        def validate_extra_vars(key, node):
+            if isinstance(node, dict):
+                return {validate_extra_vars(key, k): validate_extra_vars(k, v) for k, v in node.items()}
```
We really don't want to lose context in error messages. The `key` passed in should always be a string right? So what if you combined `k` and `key` for the value here, so instead of `validate_extra_vars(k, v)` do `validate_extra_vars(f'{key}:{k}', v)`, or whatever you'd prefer for delimiter. That would keep stacking layers of context on a rolling basis.

philipsd6 replied:
I used a '.' (dot) as the dict key delimiter, seemed appropriate for this.

AlanCoding commented on awx/main/fields.py:
```diff
@@ -858,27 +856,41 @@ def __str__(self):
+            if isinstance(node, dict):
+                return {validate_extra_vars(key, k): validate_extra_vars(k, v) for k, v in node.items()}
+            elif isinstance(node, list):
+                return [validate_extra_vars(key, x) for x in node]
```
I'd like to change `for x in node` to `for i, x in enumerate(node)` and then stack onto `key`, like `f'{key}:{i}`. It doesn't distinguish between dicts and lists... but it shouldn't matter because the user know what injectors they submitted.

philipsd6 replied:
I think the context is fully maintained now in the validation error messages. I used `key[i][i]` style for the validation error message, as between the dict dot style and this, it should be pseudo jq/jmespath style.

@AlanCoding: running `podman exec -it awx-tools_awx_1 make test_unit` works passes the new unit tests, But I'm not sure how to actually test those validation error paths.

philipsd6 committed:
Maintain nested context for validation error messages

AlanCoding commented on awx/main/fields.py:
```diff
@@ -858,27 +858,41 @@ def __str__(self):
-                        params={'value': value},
-                    )
+            if type_ == 'extra_vars':
+                validate_extra_vars(None, injector)
```
This `None` is coming through in the error message with a grammar that's not really desirable. I've actually been using the docs as an error case, because the existing example is badly formatted. That does need to be changed... but the value is:

```json
"injectors": {
    "env": {
        "THIRD_PARTY_CLOUD_API_TOKEN": "{{api_token}}"
    },
    "extra_vars": {
        "some_extra_var": "{{username}}:{{password}",
        "auth": {
            "username": "{{username}}",
            "password": "{{password}}"
        }
    }
}
```

A bit here is missing a closing `}`. The currently produced error message is:

> {'injectors': ["Syntax error rendering template for None.auth.username inside of extra_vars (unexpected '}')"]}

You know, in this line, `injector` must be a dictionary. This is fixable, but I think it will take an extra line or two of string formatting inside of `validate_extra_vars`.

I am going to write down a few validation test cases and save them for later.

AlanCoding committed:
Touchup of validation logic from testing

AlanCoding commented:
I poked at this some more and put up https://github.com/philipsd6/awx/pull/571, which is mainly concerned with error messages.
